include/unifex/async_resource.hpp

```diff
@@ -0,0 +1,423 @@
+  // 1. close all scopes
+  // 2. async destroy all resources
+  // 3. delete all containers
+  unifex::task<void> destroy() noexcept {
+    if (parent_) {
+      parent_->deregister_child(this);
+    }
+    co_await close_scope();
+    co_await destroy_resource();
+    destruct_container();
+  }
+
+  unifex::any_sender_of<> destruct(_container_base* parent) noexcept {
+    static_assert(
+        !unifex::sender_traits<decltype(destruct_impl())>::sends_done,
+        "destruct() must be unstoppable");
+    return schedule_destruct_(this, parent);
+  }
+
+  void register_child(_container_base* child) noexcept {
+    std::lock_guard lock{mutex_};
+    child->parent_ = this;
+    child->next_ = std::exchange(children_, child);
+  }
+
+  unifex::task<void> handle_construction_failure() noexcept {
+    switch (state()) {
+      case state_t::ALLOCATED: deleter_(this); break;
+      case state_t::DESTRUCT_SPAWNED: destructionEvt_.set(); break;
+      case state_t::CONSTRUCTED:
+        if (parent_ && parent_->state() == state_t::DESTROYING) {
+          co_await close_scope();
+          // bubble up cancellation, prevent returning an empty _ptr
+          co_await unifex::just_done();
+        }
+        break;
+      case state_t::DESTROYING:
+        // do nothing, ownership handed over to destroy()
+        break;
+    }
+  }
+};
+
+class async_scope_ref final {
+  unifex::v1::async_scope* scope_;
+  _container_base* container_;
+
+public:
+  /*implicit*/ async_scope_ref(unifex::v1::async_scope& scope) noexcept
+    : async_scope_ref(scope, nullptr) {}  // unmanaged scope
+
+  async_scope_ref(
+      unifex::v1::async_scope& scope, _container_base* container) noexcept
+    : scope_(&scope)
+    , container_(container) {}
+
+  void register_child(_container_base* child) noexcept {
+    UNIFEX_ASSERT(child != nullptr);
+    unifex::spawn_detached(child->destruct(container_), *scope_);
+  }
+
+  template <typename Sender>
+  [[nodiscard]] auto nest(Sender&& sender) noexcept(
+      noexcept(scope_->attach(static_cast<Sender&&>(sender))))
+      -> decltype(scope_->attach(static_cast<Sender&&>(sender))) {
+    return scope_->attach(static_cast<Sender&&>(sender));
+  }
+
+  friend bool operator==(async_scope_ref lhs, async_scope_ref rhs) noexcept {
+    return lhs.scope_ == rhs.scope_ && lhs.container_ == rhs.container_;
+  }
+
+  friend bool operator!=(async_scope_ref lhs, async_scope_ref rhs) noexcept {
+    return !(lhs == rhs);
+  }
+};
+
+template <typename Resource, typename Scheduler>
+struct _container final : _container_base {
+  _container(Scheduler scheduler) noexcept
+      : _container_base{destruct_this_resource,  //
+                        schedule_destruct,       //
+                        destroy_this_resource,   //
+                        deleter},
+        scheduler_(scheduler) {}
+
+  template <typename ResourceFactory>
+  auto construct(ResourceFactory&& factory) noexcept {
```

Comment:
This `noexcept` is probably wrong.   We need at least `std::is_nothrow_constructible_v<remove_cvref_t<ResourceFactory>, ResourceFactory>`, but that does capture the copy of `scheduler_` and whatever `on()` does.  Maybe something like
```suggestion
  auto construct(ResourceFactory&& factory) noexcept(noexcept(unifex::on(scheduler_, unifex::just_from([factory = std::forward<ResourceFactory>(factory)]() {})))) {
```